### PR TITLE
Re-apply Bech32 and TLV refactoring, this time with working note search

### PR DESCRIPTION
**Pull request by joshuatbrown**
## Issues covered
#1294

Related:
#1319
#1334 
https://github.com/planetary-social/nos/pull/1319/files#r1688461214

## Description
This is a do-over of #1319 but also fixes the bug where pasting a `note...` into the search box didn't work. See 4b88f2fa3818b56854a9feea1149ba192b9f880d for the fix.

#1319 was reverted in #1334, so this is just reapplying that work, but this time without the note search bug.

## How to test
1. Navigate to Discover
2. Paste a `note...` into search, like `note1mvcpye0rerlfqraeyyafkahvvsxrr69kxhsly9rrehkevuwllf5szvrceq`
3. Observe the note
4. Also, make sure the rest of the app still works, especially around keys and npubs and notes.

## Screenshots/Video
https://github.com/user-attachments/assets/2bc4e421-3f89-45a5-aefd-2e65d830341b

### CODE REVIEW

File: Nos/Models/PublicKey.swift [hunk 1]
```diff
@@ -37,26 +37,32 @@ struct PublicKey {
     }
     
     init?(npub: String) {
-        self.init(bech32Encoded: npub, prefix: Nostr.publicKeyPrefix)
+        self.init(bech32Encoded: npub, prefix: NostrIdentifierPrefix.publicKey)
     }
 
     init?(note: String) {
```
Comment (joshuatbrown) [inline Nos/Models/PublicKey.swift:43]: The following `init` contains the bug fix, but you know what? I'm not sure `init?(note:)` should even exist. I'm working to remove it.

Comment (joshuatbrown) [reply Nos/Models/PublicKey.swift]: It's gone!

Comment (mplorentz) [reply Nos/Models/PublicKey.swift]: 🎉 yay

**Review by mplorentz — CHANGES_REQUESTED** [review]
Hey the code looks good for me, but unfortunately tapping any link in `note1wwj68r0p84sc9te2fy7s0pxarn6zlxsnjlqmgferfararp6k8kkqh3dmd5` opens a note with an infinite spinner. I'm not seeing that on staging so I think it's something here :(

**Comment by joshuatbrown** [comment]
@mplorentz maybe an intermittent issue? I don't have that issue on that note, but I can look into it. Maybe the issue we discussed this morning -- something along the lines of #1324 but for events?

**Comment by joshuatbrown** [comment]
I guess I do see it on the Legend of Sirena link, but all the others load just fine for me.

**Comment by joshuatbrown** [comment]
@mplorentz I've double-checked and all links besides Legend of Sirena load for me both in this branch and in Staging (Legend of Sirena is broken there for me, too). I did just merge `main` into this branch, so would you mind re-testing to see if it works for you now?

**Review by mplorentz — APPROVED** [review]
It works now!
magic

**Pull request merged**